### ISSUE TITLE
Masks interfere with Precision metrics which have class_id set

### ISSUE THREAD

**Issue by forrestrice** [keras-team/keras#16271]
**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras):yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): macOS Catalina 10.15.7
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.8.0
- Python version: 3.8.2
- Bazel version (if compiling from source):
- GPU model and memory: n/a
- Exact command to reproduce: [colab](https://colab.research.google.com/drive/1yCdnTFkn2u1-1STr377OBxB1WT7L_WnA?usp=sharing)

**Describe the problem**.
If a keras model uses masking, and is fit using a `Precision` metric, then the masks are propagated to the metric computation as sample_weights. [metrics_utils](https://github.com/keras-team/keras/blob/master/keras/utils/metrics_utils.py#L617) squeezes or expands `y_pred`, `y_true`, and `sample_weight` to align the shapes. Then, if `class_id` is set on the metric, `y_true` and `y_pred` are [sliced down to that class](https://github.com/keras-team/keras/blob/master/keras/utils/metrics_utils.py#L623). However, later on when `sample_weight` is [broadcast to match](https://github.com/keras-team/keras/blob/master/keras/utils/metrics_utils.py#L678) `y_pred`, an error is thrown because it already has larger rank.

I am not totally sure if this behavior is an unintended bug, or if I am just using the metric in an unsupported way, or if I made a mistake in building my model.

**Describe the current behavior**.
An error `ValueError: weights can not be broadcast to values.` is thrown (see below for full stacktrace).

**Describe the expected behavior**.
Sample weights are squeezed or expanded in a way that handles Precision metrics with a `class_id`.

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**.

- Do you want to contribute a PR? (yes/no): No, I don't feel comfortable enough with these internals.
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):

**Standalone code to reproduce the issue**.

[colab](https://colab.research.google.com/drive/1yCdnTFkn2u1-1STr377OBxB1WT7L_WnA?usp=sharing)

**Source code / logs**.

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
[<ipython-input-2-2b7f4687dbc2>](https://localhost:8080/#) in <module>()
     17     metrics=[tf.keras.metrics.Precision(top_k=1, class_id=3)]
     18 )
---> 19 model.fit(x={"sequences": one_hot_sequences, "masks": masks}, y=one_hot_targets, epochs=3)

1 frames
[/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/func_graph.py](https://localhost:8080/#) in autograph_handler(*args, **kwargs)
   1145           except Exception as e:  # pylint:disable=broad-except
   1146             if hasattr(e, "ag_error_metadata"):
-> 1147               raise e.ag_error_metadata.to_exception(e)
   1148             else:
   1149               raise

ValueError: in user code:

    File "/usr/local/lib/python3.7/dist-packages/keras/engine/training.py", line 1021, in train_function  *
        return step_function(self, iterator)
    File "/usr/local/lib/python3.7/dist-packages/keras/engine/training.py", line 1010, in step_function  **
        outputs = model.distribute_strategy.run(run_step, args=(data,))
    File "/usr/local/lib/python3.7/dist-packages/keras/engine/training.py", line 1000, in run_step  **
        outputs = model.train_step(data)
    File "/usr/local/lib/python3.7/dist-packages/keras/engine/training.py", line 864, in train_step
        return self.compute_metrics(x, y, y_pred, sample_weight)
    File "/usr/local/lib/python3.7/dist-packages/keras/engine/training.py", line 957, in compute_metrics
        self.compiled_metrics.update_state(y, y_pred, sample_weight)
    File "/usr/local/lib/python3.7/dist-packages/keras/engine/compile_utils.py", line 459, in update_state
        metric_obj.update_state(y_t, y_p, sample_weight=mask)
    File "/usr/local/lib/python3.7/dist-packages/keras/utils/metrics_utils.py", line 70, in decorated
        update_op = update_state_fn(*args, **kwargs)
    File "/usr/local/lib/python3.7/dist-packages/keras/metrics.py", line 178, in update_state_fn
        return ag_update_state(*args, **kwargs)
    File "/usr/local/lib/python3.7/dist-packages/keras/metrics.py", line 1414, in update_state  **
        sample_weight=sample_weight)
    File "/usr/local/lib/python3.7/dist-packages/keras/utils/metrics_utils.py", line 679, in update_confusion_matrix_variables
        tf.cast(sample_weight, dtype=variable_dtype), y_pred)

    ValueError: weights can not be broadcast to values. values.rank=2. weights.rank=3. values.shape=(None, 5). weights.shape=(None, 5, 1). Received weights=Tensor("ExpandDims:0", shape=(None, 5, 1), dtype=float32), values=Tensor("strided_slice_3:0", shape=(None, 5), dtype=float32)
```

**Comment by sushreebarsa** [comment]
@sachinprasadhs I was able to replicate the issue on colab using TF v2.8.0 and tf-nightly,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/f8e255814331fc763fe4a6a723712808/untitled1.ipynb#scrollTo=TTBPZJNXktcB) .Thanks!

**Comment by sampathweb** [comment]
Thanks for reporting this issue. Its fixed now and will be available in nightly and be part of TensorFlow 2.12 release.
